include/openthread/srp_server.h

```diff
@@ -0,0 +1,212 @@
+    uint16_t            mPort;      ///< The service port.
+    uint16_t            mTxtLength; ///< The TXT record length.
+    uint8_t *           mTxtData;   ///< The standard DNS TXT record data.
+    otSrpServerHost *   mHost;      ///< The host of this service.
+    otSrpServerService *mNext;      ///< The next linked service within the same host.
+};
+
+/**
+ * This method enables/disables the SRP server.
+ *
+ * @param[in]  aInstance  A pointer to an OpenThread instance.
+ * @param[in]  aEnabled   A boolean to enable/disable the SRP server.
+ *
+ */
+void otSrpServerSetEnabled(otInstance *aInstance, bool aEnabled);
+
+/**
+ * This method sets LEASE & KEY-LEASE range that is acceptable by the SRP server.
+ *
+ * When a LEASE time is requested from a client, the granted value will be
+ * limited in range [aMinLease, aMaxLease]; and a KEY-LEASE will be granted
+ * in range [aMinKeyLease, aMaxKeyLease].
+ *
+ */
+otError otSrpServerSetLeaseRange(otInstance *aInstance,
+                                 uint32_t    aMinLease,
+                                 uint32_t    aMaxLease,
+                                 uint32_t    aMinKeyLease,
+                                 uint32_t    aMaxKeyLease);
+
+/**
+ * This method handles SRP service events by advertising the service on multicast-capable link.
+ *
+ * This function is called by the SRP server to notify a Advertising Proxy to propagate the
+ * service updates on the multiple-capable link. The Advertising Proxy should call
+ * otSrpServerHandleAdvertisingResult to return the result.
+ *
+ * @param[in]  aHost     A pointer to the otSrpServerHost object which contains the SRP updates.
+ *                       The pointer should be passed back to otSrpServerHandleAdvertisingResult, but
+ *                       the content MUST not be accessed after this method returns. The handler
+ *                       should publish/un-publish the host and each service points to this host
+ *                       with below rules:
+ *                         1. If the host has valid addresses, then it should be published or updated
+ *                            with mDNS. Otherwise, the host should be un-published (remove AAAA RRs).
+ *                         2. For each service points to this host, it must be un-published if the host
+ *                            is to be un-published. Otherwise, the handler should publish or update the
+ *                            service when it is not deleted (indicated by `otSrpServerService::mIsDeleted`)
+ *                            and un-publish it when deleted.
+ * @param[in]  aTimeout  The maximum time in milliseconds for the handler to process the service event.
+ * @param[in]  aContext  A pointer to application-specific context.
+ *
+ * @sa otSrpServerSetAdvertisingHandler
+ * @sa otSrpServerHandleAdvertisingResult
+ *
+ */
+typedef void (*otSrpServerAdvertisingHandler)(const otSrpServerHost *aHost, uint32_t aTimeout, void *aContext);
```

Comment:
I noticed that the API callback definition is mentioning "Advertising Proxy" as the intended receiver of this callback and that what the user should do in callback (e.g. "should publish update with mDNS"). I understand that this would be the common way the SRP server would be used, but thinking more abstractly, should we try to describe SRP server callback reports. Basically SRP server should be more abstract and not aware of who/what entity is using its API (i.e. whether its an Advertising Proxy or not). 

I would suggest describing what it reports and what we expect:
- This callback is notify the user that there is some update requests for a given host (`aHost`).
- The user should then process the info (check for name conflicts, etc). 
- We expect the user then to call another server API to indicate the result of its processing of the SRP update request.
- We can then give more details as example if the user is an advertising proxy (e.g., it would publish it, etc).

Does this make sense? Thoughts?

Reply:
I just renamed it from `otSrpServerUpdateHandler`(or `otSrpServerEventHandler`), but it still makes sense to me to have a more generic name.

Reply:
renamed to `otSrpServerServiceUpdateHandler`.